text/0000-decorators.md

```diff
@@ -0,0 +1,1212 @@
+
+```js
+export default class ButtonComponent extends GlimmerComponent {
+  onClick() {
+    // handle click
+  }
+}
+```
+```hbs
+<button onclick={{action this.onClick}}>Click me!</button>
+```
+
+This pattern can be used today with legacy components and controllers, but
+routes still depend heavily on `sendAction` and actions being made available
+internally via the actions object, so for the time being, we need a way to mark
+methods as actions.
+
+#### Usage and API
+
+The API for this new decorator would be much simpler than the computed API,
+since it is only used as a decorator without parameters.
+
+```ts
+// Technically `action` is a function, but we can't type it transparently that way
+const action: MethodDecorator;
+```
+
+Attempting to pass any parameters to the decorator, or to apply the decorator to
+anything other than a class method, will throw an error.
+
+## How we teach this
+
+Teaching decorators is intrinsically tied to a wider shift in the Ember
+programming model - the Ember Octane edition. From a teaching perspective, this
+edition will be completely overhauling the guides and updating all of the best
+practices as they stand. New users should see native class syntax with
+decorators as the _default_, and should not ever have to write a legacy class
+or see an example for one.
+
+With this RFC, the majority of existing examples in the Ember guides will be
+updatable to native class syntax. The exception would be examples of legacy
+components, which would be addressed separately by updating the guides to
+Glimmer Components (proposed in a separate RFC). Otherwise, all examples in the
+guides should be updated.
+
+### Updating and Interop
+
+For existing users, or users who have to interact with legacy code from a modern
+context, it'll be important to have a reference for the legacy object model.
+The current section on the object model in the guides can be moved to a legacy
+section, and a section on updating should be added. Links to relevant codemods,
+such as the
+[ember-es6-class-codemod](https://github.com/scalvert/ember-es6-class-codemod),
+should be included.
+
+## Drawbacks
+
+* The `ComputedProperty` class has long been considered intimate API. Even with
+  recent changes as part of the native getter RFC to make it more private, these
+  changes could still cause breakage.
+* The strategy for converting `computed` to decorators has one major drawback,
+  which is that decorator macros cannot easily be customized and will require a
+  bit of boilerplate in some cases. For instance, currently in Ember Decorators
+  it is possible to apply the `map` and `reduce` macros directly to a _method_,
+  which becomes the method to map or reduce by:
+
+  ```js
+  @map('array')
+  mappedArray() {}
+
+  @map('array', function() {}) mappedArray;
+  ```
+
+  With this method, only the second form would be usable. Likewise, by default
+  addons like Ember Data would need to write thin decorator wrappers around
```

Comment:
Let's make sure the Ember Data team has a chance to weigh in on this and is okay with this approach.